semantic-ast/src/Marshal/JSON.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE UndecidableInstances #-}
```

Comment:
The `UndecidableInstances` language pragma is used in `Marshal.JSON`. Without it, we see the following error:

```
semantic-ast/src/Marshal/JSON.hs:53:10: error:
    • Illegal nested constraint ‘GMarshalJSON (Rep1 t)’
      (Use UndecidableInstances to permit this)
    • In the instance declaration for ‘MarshalJSON t’
   |
53 | instance (GMarshalJSON (Rep1 t), Generic1 t) => MarshalJSON t
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

semantic-ast/src/Marshal/JSON.hs:93:10: error:
    • Illegal nested constraint ‘GMarshalJSON (Rep1 t)’
      (Use UndecidableInstances to permit this)
    • In the instance declaration for ‘GValue (Rec1 t)’
   |
93 | instance (GMarshalJSON (Rep1 t), Generic1 t) => GValue (Rec1 t) where
```

In this case, the error is coming from this instantiation of `MarshalJSON`, with a constraint. We want to create `MarshalJSON` instances for each type constructor, and ensure that type is both an instance of `Generic1` and `GMarshalJSON` such that the associated type `Rep1` can operate over the type.  My understanding is that we want to recurse into `Rep1` and we can't do that using the normal mechanics of the type system, so we need this extension? @robrix may be able to help here.

Reply:
It’s a bit more mechanical than that: the constraints available to instances are subject to certain restrictions to ensure that the constraint solver always terminates. One of those restrictions involves using fewer type variables in each constraint than appear in the instance head—because if all the constraints are “smaller” in this sense then it’s easy to verify that constraint solving will eventually run out of them.

On the other hand, when you use a type family like `Rep1 t`, there’s no way for `ghc` to prove that it won’t use `t` more than once, because that’s going to depend on the specific type which `t` is instantiated to.

Thus, this constraint violates the (admittedly fairly conservative) rules restricting instances; `UndecidableInstances` basically tells `ghc` that we’re willing to run the risk of introducing nonterminating chains of constraints and having the constraint solver infinite loop on us when we use them.

In fact, this isn’t so much of a risk—`UndecidableInstances` _sounds_ scary, but it’s not terribly difficult to show that this constraint will not be problematic so long as you don’t introduce pathological instances of `Rep1`, and since we’re deriving the `Generic1` instances (and their associated `Rep1` type instances) for all of the AST datatypes, that’s not going to happen.

Fun fact: there was a proposal to extend the compiler with a `DecidableInstances` extension to allow for an extended set of termination-checking rules which allow more instances which are safe but which `ghc`’s current checks are too conservative to allow. Sadly, it seems to have [died on the vine](https://github.com/ghc-proposals/ghc-proposals/pull/114).

Reply:
Incidentally, it’s also possible that we could work around this particular complaint of `ghc`’s by separating out the `GMarshalJSON` constraint from the application of the `Rep1` type family using a type equality constraint:

```haskell
instance (Rep1 t ~ rep, GMarshalJSON rep, Generic1 t) => GValue (Rec1 t) where …
```

Introducing a new type variable `rep`, equating it with `Rep1 t`, and constraining it to `GMarshalJSON` separately removes the nested application. I’d be curious to see whether that satisfies the constraint solver that we don’t require `UndecidableInstances`!

Reply:
mind clarifying what you mean when you point out the difference between "fewer type variables" vs. "smaller types"?

Reply:
Oh nevermind, your example with the `Show a` vs. `Show [a]` clarifies that.